pkg/inst/tests/test_utils.R

```diff
@@ -35,3 +35,64 @@ test_that("reserialize on RDD", {
   
   unlink(fileName)
 })
+
+test_that("cleanClosure on R functions", {
+  y <- c(1, 2, 3)
+  g <- function(x) { x + 1 }
+  f <- function(x) { g(x) + y }
+  newF <- cleanClosure(f)
+  env <- environment(newF)
+  expect_equal(length(ls(env)), 2)  # y, g
+  actual <- get("y", envir = env)
+  expect_equal(actual, y)
+  actual <- get("g", envir = env)
+  expect_equal(actual, g)
+  
+  # Test for nested enclosures and package variables.
+  env2 <- new.env()
+  funcEnv <- new.env(parent = env2)
+  f <- function(x) { log(g(x) + y) }
+  environment(f) <- funcEnv  # enclosing relationship: f -> funcEnv -> env2 -> .GlobalEnv
+  newF <- cleanClosure(f)
+  env <- environment(newF)
+  expect_equal(length(ls(env)), 2)  # "min" should not be included
+  actual <- get("y", envir = env)
+  expect_equal(actual, y)
+  actual <- get("g", envir = env)
+  expect_equal(actual, g)
+
+  # Test for recursive closure capture for a free variable of a function.
+  g <- function(x) { x + y }
+  f <- function(x) { lapply(x, g) + 1 }
+  newF <- cleanClosure(f)
+  env <- environment(newF)
+  expect_equal(length(ls(env)), 1)  # Only "g". "y" should be in the environemnt of g.
+  expect_equal(ls(env), "g")
+  newG <- get("g", envir = env)
+  env <- environment(newG)
+  expect_equal(length(ls(env)), 1)
+  actual <- get("y", envir = env)
+  expect_equal(actual, y)
+  
+  # Test for function (and variable) definitions.
+  f <- function(x) {
+    g <- function(y) { y * 2 }
+    g(x)
+  }
+  newF <- cleanClosure(f)
+  env <- environment(newF)
+  expect_equal(length(ls(env)), 0)  # "y" and "g" should not be included.
+  
+  # Test for overriding variables in base namespace (Issue: SparkR-196).
+  nums <- as.list(1:10)
+  rdd <- parallelize(sc, nums, 2L)
+  t = 4  # Override base::t in .GlobalEnv.
+  f <- function(x) { x > t }
+  newF <- cleanClosure(f)
+  env <- environment(newF)
+  expect_equal(ls(env), "t")
+  expect_equal(get("t", envir = env), t)
+  actual <- collect(lapply(rdd, f))
+  expected <- as.list(c(rep(FALSE, 4), rep(TRUE, 6)))
+  expect_equal(actual, expected)
+})
```

Comment:
One last request for unit test: In the old code we were explicitly filtering out broadcast variables. I think that should automatically happen now but could we add a test case which looks something like this ?
```
a <- matrix(nrow=10, ncol=10, data=rnorm(100))
aBroadcast <- broadcast(sc, a)
normMultiply <- function(x) { norm(aBroadcast.value) * x }
# Check here that cleanClosure(normMultiply) only contains `aBroadcast` and not `a`
```